### PR TITLE
GUAC-240: Implement support for JPEG server protocol and image compression.

### PR DESCRIPTION
**Pull request by flangelo**
First batch of JPEG changes. 
Added dependency on libjpeg, JPEG image compression and guacd server protocol. 
Optional code to take advantage of libjpeg-turbo feature when compressing image.

### CODE REVIEW

File: src/libguac/Makefile.am [hunk 1]
```diff
@@ -100,8 +100,8 @@ libguac_la_LDFLAGS =    \
     @PNG_LIBS@          \
     @PTHREAD_LIBS@      \
     @UUID_LIBS@         \
-    @VORBIS_LIBS@
+    @VORBIS_LIBS@	\
+    @JPEG_LIBS@
```
Comment (mike-jumper) [inline src/libguac/Makefile.am]: Mind moving `@JPEG_LIBS@` up, between `@CAIRO_LIBS@` and `@PNG_LIBS@`, such that the existing alphabetical ordering is maintained?

File: src/libguac/protocol.c [hunk 2]
```diff
@@ -74,6 +75,138 @@ ssize_t __guac_socket_write_length_double(guac_socket* socket, double d) {
...
+
+    unsigned char *mem = NULL;
+    unsigned long mem_size = 0;
+    jpeg_mem_dest(&cinfo, &mem, &mem_size);
+
+    cinfo.image_width = width; /* image width and height, in pixels */
+    cinfo.image_height = height;
+    cinfo.arith_code = TRUE;
+
+#ifdef JCS_EXTENSIONS
```
Comment (mike-jumper) [inline src/libguac/protocol.c]: This is something defined within libjpeg's headers if turbo is supported then? If so, I'm surprised (but happy!) that there is no need to add tests to `configure.ac` for this.

File: src/libguac/Makefile.am (hunk 1 above)
Comment (flangelo) [inline src/libguac/Makefile.am]: Will do

File: src/libguac/protocol.c (hunk 2 above)
Comment (flangelo) [inline src/libguac/protocol.c]: That's correct. JCS_EXTENSIONS is defined in the libjpeg-turbo headers

**Comment by mike-jumper** [comment]
Odd ... with dec7d5b, why is the PR still hanging on to 46fbf39? Ohhhh wait ... I see. Because it's a merge of dec7d5b onto the branch you're using for the PR.

You can safely do a force push with these PR's and avoid the redundant commits - GitHub will magically update.

**Comment by flangelo** [comment]
OK; re force push

File: tests/common/guac_rect.c [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    guac_common_rect_init(&max, 0, 0, 100, 100);
+    guac_common_rect_extend(&rect, &max);
+    CU_ASSERT_EQUAL(0, rect.x);
+    CU_ASSERT_EQUAL(0, rect.y);
+    CU_ASSERT_EQUAL(100, rect.width);
+    CU_ASSERT_EQUAL(100, rect.height);
+
+    /*
+     * Test adjust method
+     *  */
```
Comment (mike-jumper) [inline tests/common/guac_rect.c]:
```
     *  */
```

Oops.

**Comment by mike-jumper** [comment]
Mind if I cherry-pick and rebase as necessary to get only the `jpeg` instruction support within libguac? I think the scope of this PR is becoming a bit too broad, and I'd like to fix the strange history resulting from the merge commit.

I can do the above with a new PR internally, and then you can pick up from there by pulling from our master after it passes internal review.

We should probably deal with 21800e6 and eb67f5d via separate PRs once `guac_protocol_send_jpeg()` is finished being merged.

**Comment by flangelo** [comment]
No problem.

**Comment by mike-jumper** [comment]
OK - will continue this via #63 then. Thanks!

**Pull request closed without merging**